R/data-combined.R

```diff
@@ -589,202 +613,53 @@ DataCombined <- R6::R6Class(
-                              yOffsets = 0,
-                              xScaleFactors = 1,
-                              yScaleFactors = 1) {
+    .dataTransform = function(data) {
       # Return early if there is no data
       if (is.null(data)) {
         return(NULL)
       }
 
-      # Keep all transformation parameters and their names linked together in a
-      # data frame data structure.
-      #
-      # Additionally, if no names are provides, the transformations will apply
-      # to the entire data frame, and thus dataset names can be a placeholder for
-      # the purpose of joining of data frame with arguments and data frame with
-      # raw data that needs to be transformed.
-      dataArg <- dplyr::tibble(
-        name          = forNames %||% unique(data$name),
-        # For offsets: `0` (default for no change)
-        xOffsets      = tidyr::replace_na(xOffsets, 0),
-        yOffsets      = tidyr::replace_na(yOffsets, 0),
-        # For scale factors: `1` (default for no change)
-        xScaleFactors = tidyr::replace_na(xScaleFactors, 1),
-        yScaleFactors = tidyr::replace_na(yScaleFactors, 1)
-      )
-
-      # Update private$.dataTransformations with new transformation settings
-      private$.dataTransformations <-
-        private$.updateDataFrame(
-          dataCurrent = private$.dataTransformations,
-          dataNew     = dataArg
-        ) %>%
-        # Preserve the order from data
-        dplyr::arrange(match(name, data$name))
-
-      # Every call to method to set transformations refreshes these parameters.
-      #
-      # Thus, if there are any existing parameters from object's lifecycle,
-      # they should be removed and set with the new/default values.
-      data <-
-        data %>%
-        dplyr::select(-tidyr::any_of(colnames(private$.dataTransformations)[-1])) %>%
-        dplyr::left_join(y = private$.dataTransformations, by = "name") %>%
-        # Apply the specified transformations to the columns of interest
+      # Get transformation values
+      dataTransformations <- self$dataTransformations
+      # Transform values in data frame
+      data <- dplyr::group_by(data, name) %>%
+        # using `unique(name)` because we have grouped by the name, so it is
+        # always unique
         dplyr::mutate(
-          xValues = (xRawValues + xOffsets) * xScaleFactors,
-          yValues = (yRawValues + yOffsets) * yScaleFactors,
-          yErrorValues = yRawErrorValues * abs(yScaleFactors)
+          xValues = (xValues + dataTransformations[dataTransformations$name == unique(name), ]$xOffsets) * dataTransformations[dataTransformations$name == unique(name), ]$xScaleFactors,
+          yValues = (yValues + dataTransformations[dataTransformations$name == unique(name), ]$yOffsets) * dataTransformations[dataTransformations$name == unique(name), ]$yScaleFactors,
+          yErrorValues = yErrorValues * abs(dataTransformations[dataTransformations$name == unique(name), ]$yScaleFactors)
```

Comment:
is it worth to define `dataTransformations[dataTransformations$name == unique(name), ]` as a function?
only in 3 lines above it is used 5 times

Reply:
I don't know how this works together with `mutate`, maybe @Felixmil can help.

Reply:
not a big deal, was just a question.

Reply:
I tried some other strategy for this part of the code:

## Join approach: 
Use a join to bring values from the `self$dataTransformation` table to the original data and reference them using just their names:
```r
      data <- data %>%
        dplyr::left_join(self$dataTransformations, by = "name") %>%
        dplyr::group_by(name) %>%
        # using `unique(name)` because we have grouped by the name, so it is
        # always unique
        dplyr::mutate(
          xValues = (xValues + xOffsets) * xScaleFactors,
          yValues = (yValues + yOffsets) * yScaleFactors,
          yErrorValues = yErrorValues * abs(yScaleFactors)
        )
```

This way, the code is way easier to understand

## `{data.table}` approach

Perform a join with 
```r
      dataTransformations <- self$dataTransformations
      # Transform values in data frame
      data <-
        setDT(data)[setDT(dataTransformations),
                          `:=`(xValues = (xValues + xOffsets)*xScaleFactors,
                               yValues = (yValues + yOffsets) * yScaleFactors,
                               yErrorValues = yErrorValues * abs(yScaleFactors)),
                          on=.(name)]
      data <- tibble::as_tibble(data)
```
Here the transformation code is also easy to understand, but the `{data.table}` syntax is not.  However, as often when using `{data.table}`, this should increase performances.

## Performances
I expanded the test table from Pavel's example 1000 times, so `data` is a `1 250 000`  rows table. Here are the results:

```
Unit: milliseconds
                              expr      min        lq      mean   median       uq      max neval
      private$.dataTransform(data)  96.0567 108.04465 130.04788 116.3166 125.0462 374.7492   200
 private$.dataTransform_join(data) 132.6254 165.77815 237.44497 181.6150 376.0195 430.5048   200
   private$.dataTransform_dt(data)  42.8544  46.43925  76.90983  53.0328  62.7446 299.8599   200
```

![image](https://github.com/Open-Systems-Pharmacology/OSPSuite-R/assets/34234913/7454d795-9f51-48e7-8aaf-36fa19992413)

Shall I push the `{data.table}` implementation ? 😃

Reply:
wow.
Well, from the _performance_ point of view data.table is clearly the winner.
From the _code understanding_ point of view - @PavelBal should judge :)

And I really wonder why we have this "bimodal" distributions of the execution time in every scenario.
Just out of curiosity: if you execute the functions not 200 but e.g. 10000 times - how do the graphs look like then?

Reply:
pretty interesting